nipype/algorithms/stats.py

```diff
@@ -0,0 +1,136 @@
+    >>> seinterface.inputs.in_file = 'functional.nii'
+    >>> seinterface.inputs.label_files = 'segmentation0.nii.gz'
+    >>> seinterface.inputs.out_file = 'means.tsv'
+    >>> segments = ['CSF', 'gray', 'white']
+    >>> seinterface.inputs.class_labels = segments
+    >>> seinterface.inputs.detrend = True
+    >>> seinterface.inputs.include_global = True
+    '''
+    input_spec = SignalExtractionInputSpec
+    output_spec = SignalExtractionOutputSpec
+
+    def _run_interface(self, runtime):
+        masker, global_masker = self._process_inputs()
+
+        region_signals = masker.fit_transform(self.inputs.in_file)
+
+        if global_masker:
+            self.inputs.class_labels.insert(0, 'global')
+            global_masker.fit()
+            global_signal= global_masker.transform(self.inputs.in_file)
+            region_signals = np.hstack((global_signal, region_signals))
+
+        output = np.vstack((self.inputs.class_labels, region_signals.astype(str)))
+
+        # save output
+        np.savetxt(self.inputs.out_file, output, fmt=b'%s', delimiter='\t')
+        return runtime
+
+    def _process_inputs(self):
+        ''' validate and  process inputs into useful form '''
+
+        import nilearn.input_data as nl
+
+        # determine form of label files, choose appropriate nilearn masker
+        if len(self.inputs.label_files) > 1: # list of 3D nifti images
+            masker = nl.NiftiMapsMasker(self.inputs.label_files)
+            n_labels = len(self.inputs.label_files)
+        else: # list of size one, containing either a 3d or a 4d file
+            self.label_data = nb.load(self.inputs.label_files[0])
+            if len(self.label_data.shape) == 4: # 4d file
+                masker = nl.NiftiMapsMasker(self.label_data)
+                n_labels = self.label_data.shape[3]
+            else: # 3d file
+                if np.amax(self.label_data.get_data()) > 1: # 3d label file
+                    masker = nl.NiftiLabelsMasker(self.label_data)
+                    # assuming consecutive positive integers for regions
+                    n_labels = np.amax(self.label_data.get_data())
+                else: # most probably a single probability map for one label
+                    masker = nl.NiftiMapsMasker(self.label_data)
+                    n_labels = 1
+        masker.set_params(detrend=self.inputs.detrend)
+
+        # check label list size
+        if len(self.inputs.class_labels) != n_labels:
+            raise ValueError('The length of class_labels {} does not '
+                             'match the number of regions {} found in '
+                             'label_files {}'.format(self.inputs.class_labels,
+                                                    n_labels,
+                                                    self.inputs.label_files))
+
+        if self.inputs.include_global:
+            all_ones_mask = nb.Nifti1Image(np.ones(self._label_data_shape()), np.eye(4))
```

Comment:
This will use voxels outside of the brain! Please use the combined mask from all of the classes instead of np.ones.